terrain-manager.js

```diff
@@ -0,0 +1,180 @@
+		this.geometryUtils = geometryUtils;
+
+		this.center = new THREE.Vector3();
+		this.chunkCount = this.chunkRange * 2 + 1;
+
+		this.targetChunkIds = this._calculateTargetChunks();
+		this.currentChunks = this.targetChunkIds.map((v, i) => { return { slots: [i, i], chunkId: v } });
+
+		this.segment = 16;
+
+		/*
+		 * if following parameters are too small, memory areas of chunks can be overlaid
+		 * if too big, memory will be over allocated;
+		 */
+		this.vertexBufferSizeParam = 20;
+		this.indexBufferSizeParam = 20;
+
+		this.init();
+	}
+
+	init() {
+
+		this._generateBuffers();
+	}
+
+	_generateBuffers() {
+
+		this.bufferFactory = this.geometryUtils.generateTerrain(
+			this.chunkSize, this.chunkCount, this.segment,
+			this.vertexBufferSizeParam, this.indexBufferSizeParam
+		);
+
+		this.geometry = new THREE.BufferGeometry();
+
+		this.indexAttribute = new THREE.Uint32BufferAttribute();
+		this.indexAttribute.array = this.bufferFactory.indices;
+		this.indexAttribute.itemSize = 1;
+		this.indexAttribute.count = this.bufferFactory.indices.length;
+		this.indexAttribute.setUsage( THREE.DynamicDrawUsage );
+
+		this.positionAttribute = new THREE.Float32BufferAttribute();
+		this.positionAttribute.array = this.bufferFactory.positions;
+		this.positionAttribute.itemSize = 3;
+		this.positionAttribute.count = this.bufferFactory.positions.length / 3;
+		this.positionAttribute.setUsage( THREE.DynamicDrawUsage );
+
+		this.normalAttribute = new THREE.Float32BufferAttribute();
+		this.normalAttribute.array = this.bufferFactory.normals;
+		this.normalAttribute.itemSize = 3;
+		this.normalAttribute.count = this.bufferFactory.normals.length / 3;
+		this.normalAttribute.setUsage( THREE.DynamicDrawUsage );
+
+		this.geometry.setIndex(this.indexAttribute);
+		this.geometry.setAttribute('position', this.positionAttribute);
+		this.geometry.setAttribute('normal', this.normalAttribute);
+
+		this.geometry.clearGroups();
+
+		for (let i = 0; i < this.chunkCount * this.chunkCount; i++) {
+			this.geometry.addGroup(
+				this.bufferFactory.indexRanges[2 * i], this.bufferFactory.indexRanges[2 * i + 1], 0
+			);
+		}
+
+		this.mesh = new THREE.Mesh(
+			this.geometry, [new THREE.MeshLambertMaterial({ color: 0xff0000, wireframe: true })]
+		);
+	}
+
+	_calculateTargetChunks() {
+
+		let centerChunkGridX = Math.floor(this.center.x / this.chunkSize);
+		let centerChunkGridZ = Math.floor(this.center.z / this.chunkSize);
+
+		let targetChunks = [];
+
+		for (let i = centerChunkGridX - this.chunkRange; i < centerChunkGridX + this.chunkRange + 1; i++) {
+			for (let j = centerChunkGridZ - this.chunkRange; j < centerChunkGridZ + this.chunkRange + 1; j++) {
+				targetChunks.push(i + ':' + j);
+			}
+		}
+
+		return targetChunks;
+	}
+
+	updateCenter(pos) {
+
+		this.center = pos;
+
+		this.targetChunkIds = this._calculateTargetChunks();
+	}
+
+	updateChunk() {
+
+		const buf = this.bufferFactory;
+
+		let chunkIdToAdd = this.targetChunkIds.filter(
```

Comment:
Is this right? Only one chunk? Why not all of them that are needed?

Reply:
If we want to do one per frame then we should do an async loop that waits for frames, not do one per update call.